music_assistant/server/providers/deezer/__init__.py

```diff
@@ -0,0 +1,355 @@
+    mass: MusicAssistant,
+    instance_id: str | None = None,  # noqa: ARG001 pylint: disable=W0613
+    action: str | None = None,
+    values: dict[str, ConfigValueType] | None = None,
+) -> tuple[ConfigEntry, ...]:
+    """Return Config entries to setup this provider."""
+    # If the action is to launch oauth flow
+    if action == CONF_ACTION_AUTH:
+        # We use the AuthenticationHelper to authenticate
+        async with AuthenticationHelper(mass, values["session_id"]) as auth_helper:  # type: ignore
+            callback_url = auth_helper.callback_url
+            url = f"{DEEZER_AUTH_URL}?app_id={DEEZER_APP_ID}&redirect_uri={RELAY_URL}\
+&perms={DEEZER_PERMS}&state={callback_url}"
+            code = (await auth_helper.authenticate(url))["code"]
+            values[CONF_ACCESS_TOKEN] = await get_access_token(  # type: ignore
+                mass, DEEZER_APP_ID, DEEZER_APP_SECRET, code
+            )
+
+    return (
+        ConfigEntry(
+            key=CONF_ACCESS_TOKEN,
+            type=ConfigEntryType.SECURE_STRING,
+            label="Access token",
+            required=True,
+            action=CONF_ACTION_AUTH,
+            description="You need to authenticate on Deezer.",
+            action_label="Authenticate with Deezer",
+            value=values.get(CONF_ACCESS_TOKEN) if values else None,
+        ),
+    )
+
+
+class DeezerProvider(MusicProvider):
+    """Deezer provider support."""
+
+    client: deezer.Client
+    creds: Credential
+    _throttler: Throttler
+
+    async def handle_setup(self) -> None:
+        """Set up the Deezer provider."""
+        self._throttler = Throttler(rate_limit=4, period=1)
+        self.creds = Credential(
+            app_id=DEEZER_APP_ID,
+            app_secret=DEEZER_APP_SECRET,
+            access_token=self.config.get_value(CONF_ACCESS_TOKEN),  # type: ignore
+        )
+        try:
+            self.client = await get_deezer_client(creds=self.creds)
+        except Exception as error:
+            raise LoginFailed("Invalid login credentials") from error
+
+    @property
+    def supported_features(self) -> tuple[ProviderFeature, ...]:
+        """Return the features supported by this Provider."""
+        return SUPPORTED_FEATURES
+
+    async def search(
+        self, search_query: str, media_types=list[MediaType] | None, limit: int = 5
+    ) -> SearchResults:
+        """Perform search on music provider.
+
+        :param search_query: Search query.
+        :param media_types: A list of media_types to include. All types if None.
+        """
+        if not media_types:
+            media_types = [MediaType.ARTIST, MediaType.ALBUM, MediaType.TRACK, MediaType.PLAYLIST]
+
+        tasks = {}
+
+        async with TaskGroup() as taskgroup:
+            for media_type in media_types:
+                if media_type == MediaType.TRACK:
+                    tasks[MediaType.TRACK] = taskgroup.create_task(
+                        search_and_parse_track(
```

Comment:
shouldn't this be plural so "search_and_parse_tracks" instead of "search_and_parse_track" as there are multiple results ?

Reply:
Yes it should be. I was planning to go through all the code before i make a PR

Reply:
Fixed in newest commit